feflow/utils/hybrid_topology.py

```diff
@@ -56,23 +124,2437 @@ def __init__(self,
+                    raise AssertionError('constraint length is changing')
+
+    @staticmethod
+    def _copy_threeparticleavg(atm_map, env_atoms, vs):
+        """
+        Helper method to copy a ThreeParticleAverageSite virtual site
+        from two mapped Systems.
+
+        Parameters
+        ----------
+        atm_map : dict[int, int]
+          The atom map correspondance between the two Systems.
+        env_atoms: set[int]
+          A list of environment atoms for the target System. This
+          checks that no alchemical atoms are being tied to.
+        vs : openmm.ThreeParticleAverageSite
+
+        Returns
+        -------
+        openmm.ThreeParticleAverageSite
+        """
+        particles = {}
+        weights = {}
+        for i in range(vs.getNumParticles()):
+            particles[i] = atm_map[vs.getParticle(i)]
+            weights[i] = vs.getWeight(i)
+        if not all(i in env_atoms for i in particles.values()):
+            errmsg = ("Virtual sites bound to non-environment atoms "
+                      "are not supported")
+            raise ValueError(errmsg)
+        return openmm.ThreeParticleAverageSite(
+            particles[0], particles[1], particles[2],
+            weights[0], weights[1], weights[2],
+        )
+
+    def _handle_virtual_sites(self):
+        """
+        Ensure that all virtual sites in old and new system are copied over to
+        the hybrid system. Note that we do not support virtual sites in the
+        changing region.
+
+        TODO - remerge into a single loop
+        TODO - check that it's fine to double count here (even so, there's
+               an optimisation that could be done here...)
+        """
+        # old system
+        # Loop through virtual sites
+        for particle_idx in range(self._old_system.getNumParticles()):
+            if self._old_system.isVirtualSite(particle_idx):
+                # If it's a virtual site, make sure it is not in the unique or
+                # core atoms, since this is currently unsupported
+                hybrid_idx = self._old_to_hybrid_map[particle_idx]
+                if hybrid_idx not in self._atom_classes['environment_atoms']:
+                    errmsg = ("Virtual sites in changing residue are "
+                              "unsupported.")
+                    raise ValueError(errmsg)
+                else:
+                    virtual_site = self._old_system.getVirtualSite(
+                        particle_idx)
+                    if isinstance(
+                            virtual_site, openmm.ThreeParticleAverageSite):
+                        vs_copy = self._copy_threeparticleavg(
+                            self._old_to_hybrid_map,
+                            self._atom_classes['environment_atoms'],
+                            virtual_site,
+                        )
+                    else:
+                        errmsg = ("Unsupported VirtualSite "
```

Comment:
Testing this branch might not happen (because it's hard to find a non 3 particle site average FF), but testing the above is quite crucial because it has the oddest behaviour.